### PR TITLE
Up abstract type and `solve!`

### PR DESCRIPTION
**Pull request by tmigot**
Use new structure from SolverCore 0.3. A follow-up will be to create solver structure for the nls-variants #131

### CODE REVIEW

File: src/R2.jl [hunk 1]
```diff
@@ -51,7 +51,7 @@ stats = solve!(solver, nlp)
 "Execution stats: first-order stationary"
 ```
 """
-mutable struct R2Solver{T, V}
+mutable struct R2Solver{T, V} <: AbstractOptimizationSolver
   x::V
   gx::V
   cx::V
```
Comment (dpo) [inline src/R2.jl:57]: move stats out of here?

File: src/R2.jl [hunk 2]
```diff
@@ -73,7 +73,8 @@ end
 
 function SolverCore.solve!(
   solver::R2Solver{T, V},
-  nlp::AbstractNLPModel{T, V};
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
```
Comment (dpo) [inline src/R2.jl:77]: stats vs solver.stats

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/R2.jl (hunk 1 above)
Comment (tmigot) [reply src/R2.jl:57]: Good idea, done.

**Review by dpo — CHANGES_REQUESTED** [review]
Except for R2, the body of the solvers still don't update `stats.iter`, etc.

**Comment by tmigot** [comment]
> Except for R2, the body of the solvers still don't update `stats.iter`, etc.

Shouldn't we add this when adding the callback for each solver?

**Comment by dpo** [comment]
ah, ok.

**Review by dpo — APPROVED** [review]

**Pull request merged**